addons/sale_management/models/sale_order.py

```diff
@@ -80,6 +80,25 @@ def _check_optional_product_company_id(self):
                     bad_products=', '.join(bad_products.mapped('display_name')),
                 ))
 
+    def _get_last_sequence_domain(self, relaxed=False):
+        self.ensure_one()
+        prefix = self.sale_order_template_id.so_prefix
+        param = {'company_id': self.company_id.id, 'sale_order_template_id': self.sale_order_template_id.id or 0}
+        if not prefix:
+            # Get the latest document with no prefix: no sale_order_template_id or sale_order_template_id with empty prefix
+            where_string = "WHERE company_id = %(company_id)s AND (sale_order_template_id IS NULL OR sale_order_template_id = %(sale_order_template_id)s)"
+        else:
+            where_string = "WHERE company_id = %(company_id)s AND (sale_order_template_id = %(sale_order_template_id)s AND name LIKE %(prefix)s)"
+            param.update(prefix=f"%{prefix}%")
+        return where_string, param
```

Comment:
```suggestion
        where_string, param = super()._get_last_sequence_domain(relaxed)
        if self.sale_order_template_id:
            param['sale_order_template_id'] = self.sale_order_template_id.id
            prefix = self.sale_order_template_id.so_prefix
            if prefix:
                param['prefix'] = prefix
                where_string += " AND ..."
            else:
                where_string += " AND ..."
        return where_string, param
```

Reply:
It's not equivalent. when no template is present or when the template do not provide a prefix, we don't want to continue the sequence of prefixed sale_order. we want the default value SXXXX. What do you think of:
```py
    self.ensure_one()
    prefix = self.sale_order_template_id.so_prefix
    where_string, param = super()._get_last_sequence_domain(relaxed)
    # cast false as 0 to avoid UndefinedFunction: integer = boolean for sale_order_template_id = false
    param['sale_order_template_id'] = self.sale_order_template_id.id or 0
    if prefix:
         param['prefix'] = f"{prefix}%"
         where_string += " AND (sale_order_template_id = %(sale_order_template_id)s AND name LIKE %(prefix)s)"
    else:
        # Get the latest document with no prefix: no sale_order_template_id or sale_order_template_id with empty prefix
        where_string += " AND (sale_order_template_id IS NULL OR sale_order_template_id = %(sale_order_template_id)s)"
    return where_string, param
```

Reply:
Indeed, my solution was wrong, but this new one is still incorrect:

Imagine 2 templates without prefix:
* create one SO without template -> SO39
* We create one SO with template A --> SO40
* We create one SO with template B -> should be SO41 but it will be SO40 with your query.

I believe a join on sale_order_template is need to know if the specified template has a prefix or not :thinking: (or another solution)

Reply:
Indeed :-).
It seems that the following request is working. I added a test for different cases in sale_management.
```py
    def _get_last_sequence_domain(self, relaxed=False):
        self.ensure_one()
        prefix = self.sale_order_template_id.so_prefix
        where_string, param = super()._get_last_sequence_domain(relaxed)
        if prefix:
            param['prefix'] = f"{prefix}%"
            param['sale_order_template_id'] = self.sale_order_template_id.id
            where_string += " AND (sale_order_template_id = %(sale_order_template_id)s AND name LIKE %(prefix)s)"
        else:
            # Get any document without prefix
            where_string += " AND (sale_order_template_id IS NULL OR sale_order_template_id IN (SELECT id FROM sale_order_template WHERE so_prefix is NULL))"
        return where_string, param
```

Reply:
There'll still be the problem that if you remove the prefix on the template, then previous SO generated from this template will be considered when generating the new sequence of SO (without template/without template prefix) :confused: